File: app/src/main/java/com/mainstreetcode/teammate/fragments/main/StatEditFragment.java
package com.mainstreetcode.teammate.fragments.main;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.mainstreetcode.teammate.R;
import com.mainstreetcode.teammate.adapters.StatEditAdapter;
import com.mainstreetcode.teammate.adapters.UserAdapter;
import com.mainstreetcode.teammate.baseclasses.BaseViewHolder;
import com.mainstreetcode.teammate.baseclasses.BottomSheetController;
import com.mainstreetcode.teammate.baseclasses.HeaderedFragment;
import com.mainstreetcode.teammate.model.Stat;
import com.mainstreetcode.teammate.model.User;
import com.mainstreetcode.teammate.util.ScrollManager;
import com.mainstreetcode.teammate.viewmodel.gofers.Gofer;
import com.mainstreetcode.teammate.viewmodel.gofers.StatGofer;
import com.tunjid.androidbootstrap.view.util.InsetFlags;

import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.DiffUtil;

/**
 * Edits a Team member
 */

public class StatEditFragment extends HeaderedFragment<Stat>
        implements
        UserAdapter.AdapterListener,
        StatEditAdapter.AdapterListener {

    private static final String ARG_STAT = "stat";
    private static final int[] EXCLUDED_VIEWS = {R.id.model_list};

    private Stat stat;
    private StatGofer gofer;

    public static StatEditFragment newInstance(Stat stat) {
        StatEditFragment fragment = new StatEditFragment();
        Bundle args = new Bundle();

        args.putParcelable(ARG_STAT, stat);
        fragment.setArguments(args);
        fragment.setEnterExitTransitions();

        return fragment;
    }

    @Override
    @SuppressWarnings("ConstantConditions")
    public String getStableTag() {
        return Gofer.tag(super.getStableTag(), getArguments().getParcelable(ARG_STAT));
    }

    @Override
    @SuppressWarnings("ConstantConditions")
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        stat = getArguments().getParcelable(ARG_STAT);
        gofer = statViewModel.gofer(stat);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_headered, container, false);

        scrollManager = ScrollManager.<BaseViewHolder>with(rootView.findViewById(R.id.model_list))
                .withRefreshLayout(rootView.findViewById(R.id.refresh_layout), this::refresh)
                .withAdapter(new StatEditAdapter(gofer.getItems(), this))
                .addScrollListener((dx, dy) -> updateFabForScrollState(dy))
                .withInconsistencyHandler(this::onInconsistencyDetected)
                .withRecycledViewPool(inputRecycledViewPool())
                .withLinearLayoutManager()
                .build();

        scrollManager.getRecyclerView().requestFocus();
        return rootView;
    }

    @Override
    public void onPrepareOptionsMenu(Menu menu) {
        super.onPrepareOptionsMenu(menu);
        menu.findItem(R.id.action_delete).setVisible(gofer.canEdit() && !stat.getGame().isEnded() && !stat.isEmpty());
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_delete:
                Context context = getContext();
                if (context == null) return true;
                new AlertDialog.Builder(context).setTitle(getString(R.string.delete_stat_prompt))
                        .setPositiveButton(R.string.yes, (dialog, which) -> deleteStat())
                        .setNegativeButton(R.string.no, (dialog, which) -> dialog.dismiss())
                        .show();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onResume() {
        super.onResume();
        statViewModel.clearNotifications(stat);
    }

    @Override
    @StringRes
    protected int getFabStringResource() { return stat.isEmpty() ? R.string.stat_create : R.string.stat_update; }

    @Override
    @DrawableRes
    protected int getFabIconResource() { return R.drawable.ic_check_white_24dp; }

    @Override
    protected int getToolbarMenu() { return R.menu.fragment_stat_edit; }

    @Override protected CharSequence getToolbarTitle() {
        return getString(stat.isEmpty() ? R.string.stat_add : R.string.stat_edit);
    }

    @Override
    public InsetFlags insetFlags() {return VERTICAL;}

    @Override
    public boolean showsFab() { return !isBottomSheetShowing() && gofer.canEdit(); }

    @Override
    public int[] staticViews() {return EXCLUDED_VIEWS;}

    @Override
    protected Stat getHeaderedModel() {return stat;}

    @Override
    protected Gofer<Stat> gofer() {return gofer;}

    @Override
    protected boolean canExpandAppBar() { return false; }

    @Override
    protected void onModelUpdated(DiffUtil.DiffResult result) {
        toggleProgress(false);
        scrollManager.onDiff(result);
        viewHolder.bind(getHeaderedModel());
        Activity activity;
        if ((activity = getActivity()) != null) activity.invalidateOptionsMenu();
    }

    @Override
    public void onUserClicked(User item) {
        disposables.add(gofer.chooseUser(item).subscribe(this::onModelUpdated, defaultErrorHandler));
        hideBottomSheet();
    }

    @Override
    public void onUserClicked() {
        if (stat.getGame().isEnded()) showSnackbar(getString(R.string.stat_game_ended));
        else if (!stat.isEmpty()) showSnackbar(getString(R.string.stat_already_added));
        else pickStatUser();
    }

    @Override
    public void onTeamClicked() {
        if (stat.getGame().isEnded()) showSnackbar(getString(R.string.stat_game_ended));
        else if (!stat.isEmpty()) showSnackbar(getString(R.string.stat_already_added));
        else switchStatTeam();
    }

    @Override
    public boolean canChangeStat() { return stat.isEmpty(); }

    @Override
    public Stat getStat() { return stat; }

    @Override
    protected void onPrepComplete() {
        scrollManager.notifyDataSetChanged();
        requireActivity().invalidateOptionsMenu();
        super.onPrepComplete();
    }

    @Override
    public void onClick(View view) {
        if (view.getId() != R.id.fab) return;

        toggleProgress(true);
        disposables.add(gofer.save().subscribe(diffResult -> requireActivity().onBackPressed(), defaultErrorHandler));
    }

    private void deleteStat() {
        disposables.add(gofer.remove().subscribe(this::onStatDeleted, defaultErrorHandler));
    }

    private void onStatDeleted() {
        showSnackbar(getString(R.string.deleted_team, stat.getStatType()));
        removeEnterExitTransitions();

        Activity activity;
        if ((activity = getActivity()) == null) return;

        activity.onBackPressed();
    }

    private void pickStatUser() {
        TeamMembersFragment fragment = TeamMembersFragment.newInstance(stat.getTeam());
        fragment.setTargetFragment(this, R.id.request_stat_edit_pick);

        showBottomSheet(BottomSheetController.Args.builder()
                .setMenuRes(R.menu.empty)
                .setTitle(getString(R.string.pick_user))
                .setFragment(fragment)
                .build());
    }

    private void switchStatTeam() {
        disposables.add(gofer.switchTeams().subscribe(this::onModelUpdated, defaultErrorHandler));
    }
}

File: app/src/main/java/com/mainstreetcode/teammate/fragments/main/StatEditFragment.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Adetunji Dahunsi
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.mainstreetcode.teammate.fragments.main

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AlertDialog
import androidx.core.os.bundleOf
import androidx.recyclerview.widget.DiffUtil
import com.mainstreetcode.teammate.R
import com.mainstreetcode.teammate.adapters.StatEditAdapter
import com.mainstreetcode.teammate.adapters.UserAdapter
import com.mainstreetcode.teammate.baseclasses.BaseViewHolder
import com.mainstreetcode.teammate.baseclasses.HeaderedFragment
import com.mainstreetcode.teammate.model.Stat
import com.mainstreetcode.teammate.model.User
import com.mainstreetcode.teammate.util.ScrollManager
import com.mainstreetcode.teammate.viewmodel.gofers.Gofer
import com.mainstreetcode.teammate.viewmodel.gofers.StatGofer
import com.tunjid.androidbootstrap.view.util.InsetFlags

/**
 * Edits a Team member
 */

class StatEditFragment : HeaderedFragment<Stat>(R.layout.fragment_headered),
        UserAdapter.AdapterListener,
        StatEditAdapter.AdapterListener {

    override lateinit var headeredModel: Stat
        private set

    private lateinit var gofer: StatGofer

    override val stat: Stat get() = headeredModel

    override val insetFlags: InsetFlags get() = NO_TOP

    override val showsFab: Boolean get() = !bottomSheetDriver.isBottomSheetShowing && gofer.canEdit()

    override val stableTag: String get() = Gofer.tag(super.stableTag, arguments!!.getParcelable(ARG_STAT)!!)

    private val toolbarTitle get() = getString(if (headeredModel.isEmpty) R.string.stat_add else R.string.stat_edit)

    private val fabText get() = if (headeredModel.isEmpty) R.string.stat_create else R.string.stat_update

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        headeredModel = arguments!!.getParcelable(ARG_STAT)!!
        gofer = statViewModel.gofer(headeredModel)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        defaultUi(
                toolbarTitle = toolbarTitle,
                toolBarMenu = R.menu.fragment_stat_edit,
                fabIcon = R.drawable.ic_check_white_24dp,
                fabText = fabText,
                fabShows = showsFab
        )
        scrollManager = ScrollManager.with<BaseViewHolder<*>>(view.findViewById(R.id.model_list))
                .withRefreshLayout(view.findViewById(R.id.refresh_layout)) { this.refresh() }
                .withAdapter(StatEditAdapter(gofer.items, this))
                .addScrollListener { _, dy -> updateFabForScrollState(dy) }
                .withInconsistencyHandler(this::onInconsistencyDetected)
                .withRecycledViewPool(inputRecycledViewPool())
                .withLinearLayoutManager()
                .build().apply { recyclerView?.requestFocus() }
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        super.onPrepareOptionsMenu(menu)
        menu.findItem(R.id.action_delete)?.isVisible = gofer.canEdit() && !headeredModel.game.isEnded && !headeredModel.isEmpty
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean = when (item.itemId) {
        R.id.action_delete -> context?.run {
            AlertDialog.Builder(this).setTitle(getString(R.string.delete_stat_prompt))
                    .setPositiveButton(R.string.yes) { _, _ -> deleteStat() }
                    .setNegativeButton(R.string.no) { dialog, _ -> dialog.dismiss() }
                    .show()
            true
        } ?: true
        else -> super.onOptionsItemSelected(item)
    }

    override fun onResume() {
        super.onResume()
        statViewModel.clearNotifications(headeredModel)
    }

    override fun gofer(): Gofer<Stat> = gofer

    override fun canExpandAppBar(): Boolean = false

    override fun onModelUpdated(result: DiffUtil.DiffResult) {
        updateUi(toolbarTitle = toolbarTitle, fabText = fabText, toolbarInvalidated = true)
        transientBarDriver.toggleProgress(false)
        scrollManager.onDiff(result)
        viewHolder?.bind(headeredModel)
    }

    override fun onUserClicked(item: User) {
        disposables.add(gofer.chooseUser(item).subscribe(this::onModelUpdated, defaultErrorHandler::invoke))
        bottomSheetDriver.hideBottomSheet()
    }

    override fun onUserClicked() = when {
        headeredModel.game.isEnded -> transientBarDriver.showSnackBar(getString(R.string.stat_game_ended))
        !headeredModel.isEmpty -> transientBarDriver.showSnackBar(getString(R.string.stat_already_added))
        else -> pickStatUser()
    }

    override fun onTeamClicked() = when {
        headeredModel.game.isEnded -> transientBarDriver.showSnackBar(getString(R.string.stat_game_ended))
        !headeredModel.isEmpty -> transientBarDriver.showSnackBar(getString(R.string.stat_already_added))
        else -> switchStatTeam()
    }

    override fun canChangeStat(): Boolean = headeredModel.isEmpty

    override fun onPrepComplete() {
        scrollManager.notifyDataSetChanged()
        updateUi(toolbarInvalidated = true)
        super.onPrepComplete()
    }

    override fun onClick(view: View) {
        if (view.id != R.id.fab) return

        transientBarDriver.toggleProgress(true)
        disposables.add(gofer.save().subscribe({ requireActivity().onBackPressed() }, defaultErrorHandler::invoke))
    }

    private fun deleteStat() {
        disposables.add(gofer.remove().subscribe(this::onStatDeleted, defaultErrorHandler::invoke))
    }

    private fun onStatDeleted() {
        transientBarDriver.showSnackBar(getString(R.string.deleted_team, headeredModel.statType))
        removeEnterExitTransitions()

        activity?.onBackPressed()
    }

    private fun pickStatUser() = bottomSheetDriver.showBottomSheet {
        val caller = TeamMembersFragment.newInstance(headeredModel.team)
        caller.setTargetFragment(this@StatEditFragment, R.id.request_stat_edit_pick)

        menuRes = R.menu.empty
        title = getString(R.string.pick_user)
        fragment = caller
    }

    private fun switchStatTeam() {
        disposables.add(gofer.switchTeams().subscribe(this::onModelUpdated, defaultErrorHandler::invoke))
    }

    companion object {

        private const val ARG_STAT = "stat"

        fun newInstance(stat: Stat): StatEditFragment = StatEditFragment().apply {
            arguments = bundleOf(ARG_STAT to stat)
            setEnterExitTransitions()
        }
    }
}
